ops/lib/tcp_load_balancer/tcp_lb.py

```diff
@@ -0,0 +1,241 @@
+        self._relation_name = relation_name
+        self._member_pools = None
+        self.framework.observe(charm.on[relation_name].relation_changed, self._on_relation_changed)
+
+    def _on_relation_changed(self, event):
+        self.on.pools_changed.emit()
+
+    @property
+    def pools(self):
+        if self._member_pools is None:
+            self._member_pools = []
+            for relation in self.model.relations[self._relation_name]:
+                if not relation.units:
+                    continue
+                app_data = relation.data[relation.app]
+
+                listener_data = app_data.get('listener')
+                if listener_data is None:
+                    continue
+                listener = pickle.loads(listener_data.encode('utf-8'))
+                health_monitor_data = app_data.get('health_monitor')
+                if health_monitor_data is None:
+                    continue
+                health_monitor = pickle.loads(health_monitor_data.encode('utf-8'))
+
+                members = []
+                for unit in relation.units:
+                    member_data = relation.data[unit].get('member')
+                    if member_data is None:
+                        continue
+                    members.append(pickle.loads(member_data.encode('utf-8')))
+                self._member_pools.append(MemberPool(listener, members, health_monitor))
+        return self._member_pools
+
+
+class MemberPool:
+
+    '''A class that aggregates the information about a pool of load-balancer members'''
+
+    def __init__(self, listener, members, health_monitor=None):
+        self.members = members
+        self.listener = listener
+        self.health_monitor = health_monitor
+
+
+class LBAvailable(EventBase):
+
+    '''Event emitted by TcpLoadBalancer.on.lb_available.
+
+    This event will be emitted when a new load-balancer instance appears on a relation for a member
+    to expose its data. If a load-balancer is highly-available, there will be multiple of those
+    events fired as instances of the load-balancer are observed by the member.
+    '''
+    pass
+
+
+class TcpLoadBalancerEvents(EventSetBase):
+
+    '''Events emitted by the TcpLoadBalancer class.'''
+    lb_available = EventSource(LBAvailable)
+
+
+class TcpLoadBalancer(Object):
+
+    on = TcpLoadBalancerEvents()
+    _stored = StoredState()
+
+    def __init__(self, charm, relation_name, lb_algorithm=None):
+        super().__init__(charm, relation_name)
+        self._relation_name = relation_name
+        self._lb_algorithm = lb_algorithm
+
+        self.framework.observe(charm.on[relation_name].relation_joined, self._on_relation_joined)
+
+    def _on_relation_joined(self, event):
+        self.on.lb_available.emit()
+
+    @property
+    def _relation(self):
+        # TODO: there could be multiple independent reverse proxies in theory, address that later.
+        return self.model.get_relation(self._relation_name)
```

Comment:
This method isn't pulling its weight right now. There are only two uses, and it'll actually make the call site lazy and call this function every time it touches the attribute instead of calling it once and using a variable.